File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptPropertyParser.kt
package ca.logaritm.dezel.core

import ca.logaritm.dezel.extension.type.isNumeric
import ca.logaritm.dezel.extension.type.last2
import ca.logaritm.dezel.extension.type.last3
import ca.logaritm.dezel.extension.type.toNumber

/**
 * @class JavaScriptPropertyParser
 * @since 0.7.0
 */
public object JavaScriptPropertyParser {

	//--------------------------------------------------------------------------
	// Classes
	//--------------------------------------------------------------------------

	/**
	 * @struct Result
	 * @since 0.7.0
	 */
	public class Result(type: JavaScriptPropertyType = JavaScriptPropertyType.NULL, unit: JavaScriptPropertyUnit = JavaScriptPropertyUnit.NONE) {

		/**
		 * @property type
		 * @since 0.7.0
		 */
		public var type: JavaScriptPropertyType = JavaScriptPropertyType.NULL

		/**
		 * @property unit
		 * @since 0.7.0
		 */
		public var unit: JavaScriptPropertyUnit = JavaScriptPropertyUnit.NONE

		/**
		 * @property string
		 * @since 0.7.0
		 */
		public var string: String = ""

		/**
		 * @property number
		 * @since 0.7.0
		 */
		public var number: Double = 0.0

		/**
		 * @property boolean
		 * @since 0.7.0
		 */
		public var boolean: Boolean = false

		/**
		 * @property isNull
		 * @since 0.7.0
		 */
		public val isNull: Boolean
			get() = this.type == JavaScriptPropertyType.NULL

		/**
		 * @property isString
		 * @since 0.7.0
		 */
		public val isString: Boolean
			get() = this.type == JavaScriptPropertyType.STRING

		/**
		 * @property isNumber
		 * @since 0.7.0
		 */
		public val isNumber: Boolean
			get() = this.type == JavaScriptPropertyType.NUMBER

		/**
		 * @property isBoolean
		 * @since 0.7.0
		 */
		public val isBoolean: Boolean
			get() = this.type == JavaScriptPropertyType.BOOLEAN

		/**
		 * @constructor
		 * @since 0.7.0
		 */
		init {
			this.type = type
			this.unit = unit
		}

		/**
		 * @constructor
		 * @since 0.7.0
		 */
		public constructor(string: String) : this(JavaScriptPropertyType.STRING, JavaScriptPropertyUnit.NONE) {
			this.string = string
		}

		/**
		 * @constructor
		 * @since 0.7.0
		 */
		public constructor(number: Double) : this(JavaScriptPropertyType.NUMBER, JavaScriptPropertyUnit.NONE) {
			this.number = number
		}

		/**
		 * @constructor
		 * @since 0.7.0
		 */
		public constructor(number: Double, unit: JavaScriptPropertyUnit) : this(JavaScriptPropertyType.NUMBER, unit) {
			this.number = number
		}

		/**
		 * @constructor
		 * @since 0.7.0
		 */
		public constructor(boolean: Boolean) : this(JavaScriptPropertyType.BOOLEAN, JavaScriptPropertyUnit.NONE) {
			this.boolean = boolean
		}
	}

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @method parse
	 * @since 0.7.0
	 */
	public fun parse(value: JavaScriptValue): Result? {

		when (true) {
			value.isNull      -> return Result()
			value.isUndefined -> return Result()
			value.isString    -> return this.parse(value.string)
			value.isNumber    -> return Result(value.number)
			value.isBoolean   -> return Result(value.boolean)
			else              -> {}
		}

		return null
	}

	/**
	 * @method parse
	 * @since 0.7.0
	 */
	public fun parse(value: String): Result? {

		if (value.isNumeric) {

			if (value.takeLast(1) == "%") {
				return Result(value.toNumber(), JavaScriptPropertyUnit.PC)
			}

			when (value.last2) {
				"px" -> return Result(value.toNumber(), JavaScriptPropertyUnit.PX)
				"vw" -> return Result(value.toNumber(), JavaScriptPropertyUnit.VW)
				"vh" -> return Result(value.toNumber(), JavaScriptPropertyUnit.VH)
				"pw" -> return Result(value.toNumber(), JavaScriptPropertyUnit.PW)
				"ph" -> return Result(value.toNumber(), JavaScriptPropertyUnit.PH)
				"cw" -> return Result(value.toNumber(), JavaScriptPropertyUnit.CW)
				"ch" -> return Result(value.toNumber(), JavaScriptPropertyUnit.CH)
				else -> {}
			}

			when (value.last3) {
				"deg" -> return Result(value.toNumber(), JavaScriptPropertyUnit.DEG)
				"rad" -> return Result(value.toNumber(), JavaScriptPropertyUnit.RAD)
				else  -> {}
			}

			return Result(value.toNumber())

		} else if (value == "null") {

			return Result()

		} else if (value == "true") {

			return Result(true)

		} else if (value == "false") {

			return Result(false)

		}

		return null
	}
}

File: platforms/ios/Dezel/Dezel/Core/JavaScriptPropertyParser.swift
/**
 * Parses numeric properties.
 * @class JavaScriptPropertyParser
 * @since 0.7.0
 */
open class JavaScriptPropertyParser {

	//--------------------------------------------------------------------------
	// MARK: Classes
	//--------------------------------------------------------------------------

	/**
	 * The result.
	 * @struct Result
	 * @since 0.7.0
	 */
	public class Result {

		/**
		 * The result's type.
		 * @property type
		 * @since 0.7.0
		 */
		public var type: JavaScriptPropertyType = .null

		/**
		 * The result's unit.
		 * @property unit
		 * @since 0.7.0
		 */
		public var unit: JavaScriptPropertyUnit = .none

		/**
		 * The result's string.
		 * @property string
		 * @since 0.7.0
		 */
		public var string: String!

		/**
		 * The result's number.
		 * @property number
		 * @since 0.7.0
		 */
		public var number: Double!

		/**
		 * The result's boolean.
		 * @property boolean
		 * @since 0.7.0
		 */
		public var boolean: Bool!

		/**
		 * Indicate whether the result is null.
		 * @property isNull
		 * @since 0.7.0
		 */
		public var isNull: Bool {
			return self.type == .null
		}

		/**
		 * Indicate whether the result is a string.
		 * @property isString
		 * @since 0.7.0
		 */
		public var isString: Bool {
			return self.type == .string
		}

		/**
		 * Indicate whether the result is a number.
		 * @property isNumber
		 * @since 0.7.0
		 */
		public var isNumber: Bool {
			return self.type == .number
		}

		/**
		 * Indicate whether the result is a boolean.
		 * @property isBoolean
		 * @since 0.7.0
		 */
		public var isBoolean: Bool {
			return self.type == .boolean
		}

		/**
		 * Initializes the result.
		 * @constructor
		 * @since 0.7.0
		 */
		public init() {
			self.type = .null
			self.unit = .none
		}

		/**
		 * Initializes the result.
		 * @constructor
		 * @since 0.7.0
		 */
		public init(string: String) {
			self.type = .string
			self.unit = .none
			self.string = string
		}

		/**
		 * Initializes the result.
		 * @constructor
		 * @since 0.7.0
		 */
		public init(number: Double) {
			self.type = .number
			self.unit = .none
			self.number = number
		}

		/**
		 * Initializes the result.
		 * @constructor
		 * @since 0.7.0
		 */
		public init(number: Double, unit: JavaScriptPropertyUnit) {
			self.type = .number
			self.unit = unit
			self.number = number
		}

		/**
		 * Initializes the result.
		 * @constructor
		 * @since 0.7.0
		 */
		public init(boolean: Bool) {
			self.type = .boolean
			self.unit = .none
			self.boolean = boolean
		}
	}

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * Parses the specified JavaScript value.
	 * @method parse
	 * @since 0.7.0
	 */
	public static func parse(_ value: JavaScriptValue) -> Result? {

		switch (true) {

			case value.isNull:
				return Result()
			case value.isUndefined:
				return Result()
			case value.isString:
				return self.parse(value.string)
			case value.isNumber:
				return Result(number: value.number)
			case value.isBoolean:
				return Result(boolean: value.boolean)

			default:
				break
		}

		return nil
	}

	/**
	 * Parses the specified string.
	 * @method parse
	 * @since 0.7.0
	 */
	public static func parse(_ value: String) -> Result? {

		if (value.isNumeric) {

			if (value.suffix(1) == "%") {
				return Result(number: value.toNumber(), unit: .pc)
			}

			switch (value.last2) {

				case "px": return Result(number: value.toNumber(), unit: .px)
				case "vw": return Result(number: value.toNumber(), unit: .vw)
				case "vh": return Result(number: value.toNumber(), unit: .vh)
				case "pw": return Result(number: value.toNumber(), unit: .pw)
				case "ph": return Result(number: value.toNumber(), unit: .ph)
				case "cw": return Result(number: value.toNumber(), unit: .cw)
				case "ch": return Result(number: value.toNumber(), unit: .ch)

				default:
					break
			}

			switch (value.last3) {

				case "deg": return Result(number: value.toNumber(), unit: .deg)
				case "rad": return Result(number: value.toNumber(), unit: .rad)

				default:
					break
			}

			return Result(number: value.toNumber())

		} else if (value == "null") {

			return Result()

		} else if (value == "true") {

			return Result(boolean: true)

		} else if (value == "false") {

			return Result(boolean: false)

		}

		return nil
	}
}
